[nodelabeller, virt-handler]: Remove nodelabeller capabilities package and decouple vmcontroller from nodelabeller

Acedus opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. Consider creating this PR as draft: https://github.com/kubevirt/kubevirt/blob/main/CONTRIBUTING.md#consider-opening-your-pull-request-as-draft
2. Follow the instructions for writing a release note from k8s: https://git.k8s.io/community/contributors/guide/release-notes.md
-->

### What this PR does
**Before this PR:**

nodelabeller currently offers an additional package under `api` called `capabilities`, what it does is provide an exported structure that mimics a host `capabilities.xml`, allowing data to be unmarshalled into it and subsequently utilized by both nodelabeller (which currently only uses the host CPU counter) and the vmcontroller.

This creates a dependency between the vmcontroller and nodelabeller, as the former relies on the capabilities package for loading the capabilities.

Additionally, this setup impacts testability, as node-labeller uses an internal function called loadHostCapabilities, which requires reading a file to load the `capabilities.xml`, even though most of its contents are not used and could be easily substituted with a DI during the nodelabeller construction.

**After this PR:**

The capabilities package is replaced by the [https://pkg.go.dev/libvirt.org/go/libvirtxml](libvirt-go-xml-module) package. This package:
1. Is already included as a Kubevirt dependency and is extensively used as part of [virtwrap/convert.go](https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-launcher/virtwrap/libvirtxml/convert.go).
2. Serves the same purpose as the former `nodelabeller/capabilities` package.

Moreover, this PR decouples the vmcontroller and nodelabeller by shifting the capabilities inference to be part of the virt-handler initialization. The relevant fields are now determined and are passed to both components according to their respective needs.

<!-- (optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*: -->

Partially addresses https://github.com/kubevirt/kubevirt/issues/12727

### Special notes for your reviewer

This PR is dependent on https://github.com/kubevirt/kubevirt/pull/12761

The main goal of the PR is to enhance the testability of nodelabeller. As such, subsequent PRs will be created to address and remove problematic logic (namely loadAll()).

<!-- optional -->

### Checklist

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

### Release note
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
4. If no release note is required, just write "NONE".
-->
```release-note
NONE
```

Acedus commented:
/cc @orelmisan 
/cc @Barakmor1

Acedus commented:
[Change](https://github.com/kubevirt/kubevirt/compare/551b9e9f59c3e49705583f57017ff3c2d891efcd..2993085b02639e3b6863cf064c16e86684e743bb) - Removed `pkg/virt-handler/node-labeller/testdata/capabilities_with_numa.xml` as it was no longer used.

alicefr commented:
Nice!
fyi @victortoso

alicefr commented on cmd/virt-handler/virt-handler.go:
```diff
@@ -306,14 +305,24 @@ func (app *virtHandlerApp) Run() {
+	if err != nil {
+		panic(err)
+	}
+	err = capabilities.Unmarshal(string(rawFile))
```
nit: inline

Acedus replied:
Done.

alicefr commented:
The code looks very nice,thanks!
/lgtm

Acedus commented:
[Change](https://github.com/kubevirt/kubevirt/compare/2993085b02639e3b6863cf064c16e86684e743bb..144785d1fa10df80e99b3318892b9d1d69b20086) - Inlined error handling for capabilities unmarshaling.

orelmisan commented on pkg/virt-handler/node-labeller/node_labeller.go:
```diff
@@ -71,19 +71,19 @@ type NodeLabeller struct {
 
-func NewNodeLabeller(clusterConfig *virtconfig.ClusterConfig, nodeClient k8scli.NodeInterface, host string, recorder record.EventRecorder) (*NodeLabeller, error) {
-	return newNodeLabeller(clusterConfig, nodeClient, host, nodeLabellerVolumePath, recorder)
+func NewNodeLabeller(clusterConfig *virtconfig.ClusterConfig, nodeClient k8scli.NodeInterface, host string, cpuCounter *libvirtxml.CapsHostCPUCounter, recorder record.EventRecorder) (*NodeLabeller, error) {
```
nit: Please consider adding the `cpuCounter` as the last parameter.
It is reasonable to assume that follow-up PRs will extract additional file-system reads and add additional parameters here (or a builder).

Acedus replied:
Done.

orelmisan commented on pkg/virt-handler/node-labeller/node_labeller_test.go:
```diff
@@ -51,9 +52,14 @@ var _ = Describe("Node-labeller ", func() {
 		config, _, _ := testutils.NewFakeClusterConfigUsingKV(kubevirt)
 		recorder := record.NewFakeRecorder(100)
 		recorder.IncludeObject = true
+		cpuCounter := &libvirtxml.CapsHostCPUCounter{
+			Name:      "tsc",
+			Frequency: 4008012000,
+			Scaling:   "no",
+		}
```
It could be nice to add a test for it.

Acedus replied:
It's already being covered by `cpu_plugin_test.go`, theoretically the cpuCounter value could be nil for this test and it wouldn't matter. With that said, most of the existing tests for nodelabeller could be improved upon, but I think that would be better saved for another PR.

Acedus replied:
After talking about it offline we've agreed that a test should be added as part of the nodelabeller logic.

orelmisan commented on pkg/virt-handler/options.go:
```diff
@@ -119,3 +122,16 @@ func disksInfoToDisksInfo(disksInfo map[string]*containerdisk.DiskInfo) map[stri
 	return info
 }
+
+func convertSiblings(siblings string) []uint32 {
```
Is this logic covered by tests?

Acedus replied:
So `options.go` in its entirety doesn't have tests AFAIK.

What could be done is add capabilities to `vm_test.go` controller creation (it's currently passing nil for capabilities) and then it would be covered by the call to `Execute` during tests. 

I'm a bit cautious about adding this though, let me know what you think.

orelmisan replied:
I think we could add `options_test.go` and test at least `capabilitiesToTopology` in order to verify that the change introduced by this PR hasn't caused a regression.

orelmisan reviewed (commented):
Thank you for the PR @Acedus.
This is a good first step towards better testability of `NodeLabeller`.

Acedus commented on pkg/virt-handler/node-labeller/node_labeller.go:
```diff
@@ -249,11 +249,11 @@ func (n *NodeLabeller) prepareLabels(node *v1.Node, cpuModels []string, cpuFeatu
-	} else if err != nil {
-		n.logger.Reason(err).Error("failed to get tsc cpu frequency, will continue without the tsc frequency label")
+	} else {
+		n.logger.Error("failed to get tsc cpu frequency, will continue without the tsc frequency label")
```
I think it might be worth to remove the log entry in its entirety, it kind of is a precedence (save for `isNodeRealtimeCapable` which should be taken care of as part of https://github.com/kubevirt/kubevirt/pull/12522.

orelmisan replied:
>  (save for isNodeRealtimeCapable which should be taken care of as part of #12522).

I don't follow this part, could you please elaborate?

Acedus commented on pkg/virt-launcher/virtwrap/converter/vcpu/vcpu.go:
```diff
@@ -163,9 +163,9 @@ func cpuChunksToCells(cpuSet []int, nodeTopology *v1.Topology) (cores [][][]uint
 				for _, thread := range cpu.Siblings {
-					visited[thread] = struct{}{}
+					visited[uint32(thread)] = struct{}{}
 					if _, exists := threads[int(thread)]; exists {
-						core = append(core, thread)
+						core = append(core, uint32(thread))
```
The logic of vcpu.go already conforms to `uint32` (as per the precedent set by the `virtwrap/api`). Changing it to fit the new `thread` type would require far more substantial changes to the schema (which may be desired, but out of scope for this PR).

Acedus replied:
Scrap that, it was my mistake, probably when I was messing around with the cmd/v1 proto generation. Removing that.

Acedus commented on cmd/virt-handler/virt-handler.go:
```diff
@@ -306,14 +306,23 @@ func (app *virtHandlerApp) Run() {
 	stop := make(chan struct{})
 	defer close(stop)
-	var capabilities *libvirtxml.Caps
+	capabilities := &libvirtxml.Caps{}
```
Done.

Acedus commented on cmd/virt-handler/virt-handler.go:
```diff
@@ -306,14 +306,23 @@ func (app *virtHandlerApp) Run() {
 	var hostCpuModel string
+
+	capsFile := filepath.Join(nodelabeller.NodeLabellerVolumePath, "capabilities.xml")
+	rawFile, err := os.ReadFile(capsFile)
```
Done, also changed `rawFile` name to be more meaningful (now `hostCapsFile`).

Acedus commented on pkg/virt-handler/node-labeller/cpu_plugin_test.go:
```diff
@@ -83,7 +83,7 @@ var _ = Describe("Node-labeller config", func() {
 
 		Expect(cpuFeatures).To(HaveLen(4), "number of features must match")
-		counter, err := nlController.capabilities.GetTSCCounter()
+		counter := nlController.GetTSCCounter()
```
Done.

orelmisan replied:
Still not removed :(

Acedus replied:
Done, was previously done in the third commit.

Acedus commented:
[Change](https://github.com/kubevirt/kubevirt/compare/53c9b4ba7dce21a7abb356ce400888f3607e9d28..ea4cd6051483d76843c1d4de6ec7ac895dadf059) - Removed redundant uint32 casting from `vcpu.go`, added TSC label test to `node_labeller_test.go`.

Acedus commented:
[Change](https://github.com/kubevirt/kubevirt/compare/ea4cd6051483d76843c1d4de6ec7ac895dadf059..270a8719696d0076ca010be38529d2c1a39cc5db) - Amend commit message to refer to the PR instead of the commit that introduced libvirtxml.

orelmisan commented on cmd/virt-handler/virt-handler.go:
```diff
@@ -305,14 +305,28 @@ func (app *virtHandlerApp) Run() {
 
 	stop := make(chan struct{})
 	defer close(stop)
-	var capabilities *libvirtxml.Caps
```
Please consider adding a few words to the commit message about the two major advantages we get from this change:
1. Improved testability of NodeLabeller 
2. Decoupling of VMController from NodeLabeller.

Acedus replied:
> I don't follow this part, could you please elaborate?

Logging the failure to obtain TSC frequency values for every `prepareLabels` run is redundant. It is possible to move the log message somewhere else (e.g. node-labeller `Run`), and upon further consideration I'll do just that.

> I think we could add options_test.go and test at least capabilitiesToTopology in order to verify that the change introduced by this PR hasn't caused a regression.

I've added it in a separate commit, the construction of the capabilities for the test is a bit complex but I preferred it over adding a testdata file or declaring a variable that contains the relevant data in XML form, would like feedback on that.

> Please consider adding a few words to the commit message about the two major advantages we get from this change:
> 1. Improved testability of NodeLabeller
> 2. Decoupling of VMController from NodeLabeller.

Done.

orelmisan commented on pkg/virt-handler/node-labeller/api/BUILD.bazel:
```diff
@@ -1,9 +0,0 @@
-load("@io_bazel_rules_go//go:def.bzl", "go_library")
```
Is it possible to perform these removals in the first commit?
Since the replacement happens there.

Acedus replied:
Good catch, done.

orelmisan commented on pkg/virt-handler/node-labeller/node_labeller_test.go:
```diff
@@ -162,6 +168,17 @@ var _ = Describe("Node-labeller ", func() {
+
+		node := retrieveNode(kubeClient)
+		Expect(node.Labels).To(SatisfyAll(
+			HaveKey(v1.CPUTimerLabel+"tsc-frequency"),
+			HaveKey(v1.CPUTimerLabel+"tsc-scalable"),
```
Please also assert the values.

Acedus replied:
Done.

orelmisan commented on pkg/virt-handler/node-labeller/cpu_plugin_test.go:
```diff
@@ -74,20 +82,15 @@ var _ = Describe("Node-labeller config", func() {
-		Expect(counter).ToNot(BeNil())
-		Expect(counter.Frequency).To(BeNumerically("==", 4008012000))
 
+		Expect(nlController.cpuCounter).ToNot(BeNil())
+		Expect(nlController.cpuCounter.Name).To(Equal("tsc"))
+		Expect(nlController.cpuCounter.Frequency).To(BeNumerically("==", 4008012000))
```
Since there is a better test under `node_labeller_test.go`, I'm not sure we need to keep this one as well.
It asserts the internal state of `nlController` (which can break) while the other test tests its observed behavior.

Acedus replied:
That's fair, I kept it since the original test had it but now that we have a better one I agree that it's unnecessary.

vladikr commented:
In general, the direction looks good to me.
/approve

Please take a look at the failing tests.

iholder101 commented on pkg/virt-handler/options.go:
```diff
@@ -119,3 +122,16 @@ func disksInfoToDisksInfo(disksInfo map[string]*containerdisk.DiskInfo) map[stri
 	return info
 }
+
+func convertSiblings(siblings string) []uint32 {
```
I think the function name can be improved as it's not very informative.
WDYT about something like `convertListOfIntStringToSlice()`?

Acedus replied:
I agree, done.

iholder101 reviewed (commented):
The direction looks great to me.
See minor comments below.

Acedus commented on pkg/virt-handler/node-labeller/cpu_plugin.go:
```diff
@@ -150,8 +151,12 @@ func (n *NodeLabeller) loadHostSupportedFeatures() error {
+	if err != nil {
+		return err
+	}
+	err = n.capabilities.Unmarshal(string(rawFile))
```
To use the `Unmarshal` method of libvirtxml `Caps` would require me to alter `getStructureFromXMLFile` in a way that would obstruct its other uses (e.g. `getDomCapabilities`).

I think the general direction for `cpu_plugin.go` as a whole is towards complete deprecation in favor of construction using basic types (the necessary ones) and inference at the virt-handler level. Should it prove its use, it may be worth it to have `cpu_plugin.go` as its own separate package with the responsibility of obtaining said information.

Acedus commented:
[Change](https://github.com/kubevirt/kubevirt/compare/270a8719696d0076ca010be38529d2c1a39cc5db..134a8714c8ed028e0e75ba4005e58f3b0b529362) - Addressed review notes, added `options_test.go` to test VMI options parsing functionality, adjusted commit messages to better reflect changes.

Acedus commented:
[Change](https://github.com/kubevirt/kubevirt/compare/134a8714c8ed028e0e75ba4005e58f3b0b529362..dd4d4cb5aef8d97718eeb24d9b3ef13bb953fe94) - Made `options_test.go` more modular.

Acedus commented on pkg/virt-handler/options_test.go:
Complex setup can be:
1. Logically separated into functions.
2. Be declared as an XML string variable and unmarshaled.
3. Mimic previous behavior by reading a testdata file and unmarshaling it. 

WDYT?

Acedus commented:
[Change](https://github.com/kubevirt/kubevirt/compare/dd4d4cb5aef8d97718eeb24d9b3ef13bb953fe94..230ad2c11b31f32bfaf182a0c03f3c01fc912155) - Fixed issue where tsc-scaling label was using yes/no instead of boolean value due to how libvirtxml works.

Acedus commented:
[Change](https://github.com/kubevirt/kubevirt/compare/230ad2c11b31f32bfaf182a0c03f3c01fc912155..63c15e0a8c9738279e945bd5b99428cda0054343) - Use %t instead of %v for boolean value.

Acedus commented:
/retest pull-kubevirt-e2e-k8s-1.31-sig-compute

Acedus commented:
/test pull-kubevirt-e2e-k8s-1.31-sig-compute

iholder101 commented on pkg/virt-handler/options_test.go:
```diff
@@ -0,0 +1,171 @@
+	})
+})
+
+func matchNumaCells(expectedCells []*cmdv1.Cell) []OmegaMatcher {
+	matchers := make([]OmegaMatcher, len(expectedCells))
```
Pretty cool to use omega matchers like that!

iholder101 reviewed (commented):
Thanks a lot for this PR! I really like the direction you're going for.

As I've suggested in the mailing list before, I do suggest to create a design proposal. While it's not a must, I think it'll help people understand the overall plan you have in mind, therefore reviews would likely be faster and easier.

Acedus commented on pkg/virt-handler/node-labeller/node_labeller_test.go:
```diff
@@ -162,6 +168,17 @@ var _ = Describe("Node-labeller ", func() {
+
+		node := retrieveNode(kubeClient)
+		Expect(node.Labels).To(SatisfyAll(
+			HaveKeyWithValue(v1.CPUTimerLabel+"tsc-frequency", "4008012000"),
+			HaveKeyWithValue(v1.CPUTimerLabel+"tsc-scalable", "false"),
```
Ideally I'd like to avoid using unexported fields in test (and eventually change the package to nodelabeller_test), ergo the explicit values. I can create a getter for CPU counter if you prefer that style.

iholder101 replied:
> Ideally I'd like to avoid using unexported fields in test and eventually change the package to nodelabeller_test)

Even if this test was in another package, you've added `cpuCounter` as an argument for `NewNodeLabeller()`. In this test you're defining a `cpuCounter` struct with explicitly set values, then you're passing it to this constructor. So I don't really see how my suggestion uses unexported fields.

Acedus replied:
Perhaps I misunderstood your suggestion, the scope of `cpuCounter` remains limited to the initialization function in which it is declared, therefore accessing it from `nlController` later on is equivalent to accessing an unexported field. 

I can either declare `cpuCounter` in the outer scope or introduce a getter method for the respective field in node-labeller, WDYT?

iholder101 replied:
Yeah I see your point.
If this value were to change sometime in the future, it'd be better to not change it in two places, so I think I would move `cpuCounter` out of the init function, make it global, and assign it in the BeforeEach block. But that's not a must from my end :)

Thanks for clarifying!

Acedus replied:
Ack, done.

Acedus commented on pkg/virt-handler/node-labeller/cpu_plugin_test.go:
```diff
@@ -74,20 +82,15 @@ var _ = Describe("Node-labeller config", func() {
-		Expect(counter).ToNot(BeNil())
-		Expect(counter.Frequency).To(BeNumerically("==", 4008012000))
 
+		Expect(nlController.cpuCounter).ToNot(BeNil())
+		Expect(nlController.cpuCounter.Name).To(Equal("tsc"))
+		Expect(nlController.cpuCounter.Frequency).To(BeNumerically("==", 4008012000))
```
My vision for `cpu_plugin_test.go` in general is to be removed, but I'll remove this test regardless as you're right, it's redundant.

Acedus commented on pkg/virt-handler/node-labeller/node_labeller.go:
```diff
@@ -340,3 +333,7 @@ func (n *NodeLabeller) alertIfHostModelIsObsolete(originalNode *v1.Node, hostMod
 	return nil
 }
+
+func (n *NodeLabeller) hasTSCCounter() bool {
+	return n.cpuCounter.Name == "tsc"
+}
```
I do wonder that as well, I couldn't find a reference as to why it was decided to be implemented as such but the official libvirtxml package unmarshals it to a single object instead of a slice.

iholder101 replied:
Thank you!

Acedus commented:
[Change](https://github.com/kubevirt/kubevirt/compare/63c15e0a8c9738279e945bd5b99428cda0054343..7c9acf26cb98a149bcc057b7a3cf2431ac581869) - Rewrote `options_test.go` and now introduces it in the first commit to ensure that tests pass with the previous API (nodelabeller/capabilities), removed redundant TSC value assertion as part of `cpu_plugin_test.go`

Acedus commented on cmd/virt-handler/virt-handler.go:
```diff
@@ -306,14 +305,28 @@ func (app *virtHandlerApp) Run() {
 	var hostCpuModel string
+
+	hostCapsFile, err := os.ReadFile(filepath.Join(nodelabeller.NodeLabellerVolumePath, "capabilities.xml"))
+	if err != nil {
+		panic(err)
+	}
+
+	if err := capabilities.Unmarshal(string(hostCapsFile)); err != nil {
+		panic(err)
+	}
+
 	nodeLabellerrecorder := broadcaster.NewRecorder(scheme.Scheme, k8sv1.EventSource{Component: "node-labeller", Host: app.HostOverride})
-	nodeLabellerController, err := nodelabeller.NewNodeLabeller(app.clusterConfig, app.virtCli.CoreV1().Nodes(), app.HostOverride, nodeLabellerrecorder)
+	nodeLabellerController, err := nodelabeller.NewNodeLabeller(app.clusterConfig,
+		app.virtCli.CoreV1().Nodes(),
+		app.HostOverride,
+		nodeLabellerrecorder,
+		capabilities.Host.CPU.Counter,
+	)
```
> I think it would be cleaner if a separate "unit" would be in charge of the inference (package / function) instead of doing so in the Run() function which is already pretty long and non-cohesive. Bear in mind that in the future more inference code will be added (e.g. for dom capabilities, host supported features, etc).

I agree, I wasn't sure whether to introduce it during this PR or not but we can do that.
My idea was for a new package under virt-handler (e.g. `NodeCapabilityManager`) and slowly migrate `cpu_plugin.go` logic into it WDYT?

> It's confusing that the mount path and even the init container's script is still called node-labeller. Perhaps we can rename all of these to more informative names. I think it's important since then it will be clearer of what your long-term plan is.

I agree, I'll add it in a separate commit.

> Further to the above, I think that nodelabeller.NodeLabellerVolumePath should now reside outside of node-labeller (with a different name).

Relevant to the previous comment, will be introduced as part of that commit.

iholder101 replied:
> I agree, I wasn't sure whether to introduce it during this PR or not but we can do that.

The direction sounds good to me.

> My idea was for a new package under virt-handler (e.g. NodeCapabilityManager) and slowly migrate cpu_plugin.go logic into it WDYT?

A general note about future plans: we can either review this PR individually or we need to discuss the overall plan and review your PRs with more context. As suggested in other places, I think a design proposal would be beneficial in this case, as it will allow you to work with smaller PRs and ease the review process.

Acedus replied:
As I've responded in the tracker issue https://github.com/kubevirt/kubevirt/issues/12727, I am currently working on a design proposal to better discuss the overhaul in its entirety. 

For the time being, I believe this PR, even in its standalone form, is essential for any future changes pertaining to the node-labeller.

iholder101 replied:
> As I've responded in the tracker issue https://github.com/kubevirt/kubevirt/issues/12727, I am currently working on a design proposal to better discuss the overhaul in its entirety.

Happy to hear that!
Sorry for missing your response.

> For the time being, I believe this PR, even in its standalone form, is essential for any future changes pertaining to the node-labeller.

I agree, introducing the design proposal shouldn't block this PR.
Thanks!

Acedus commented:
[Change](https://github.com/kubevirt/kubevirt/pull/12652/commits/11720a3a7c934222982cffaa27eb6e8d6c76c70f) - Added commit that moves the `node-labeller.sh` script out of `node-labeller` and renames it.

Barakmor1 commented on pkg/virt-handler/options_test.go:
```diff
@@ -0,0 +1,173 @@
+			Entry("Single cell, single page, single CPU per cell", 1, 1, 1),
+			Entry("Multiple cells, multiple pages, multiple CPUs per cell", 4, 3, 4),
+		)
+	})
```
I’m a bit unsure about the process of writing this test in one commit and then modifying it in another. It’s making the review a bit challenging for me.

Acedus replied:
I understand, however, it was done to show that previous logic was functioning properly so that the new logic wouldn't seem too abrupt of an introduction. I can remove it should it be required, but the code itself doesn't change much between the two commits.

Barakmor1 replied:
Thank you for explaining. From a review perspective, if future commits modify code already changed in this PR, I’ll need to review it again. This makes reviewing commit-by-commit a bit challenging.

Barakmor1 commented:
@Acedus In general this PR look like a great improvement to kubevirt, i'll give a deeper look soon.

Have you considered about doing the same in a follow up PR for [getDomCapabilities](https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-handler/node-labeller/cpu_plugin.go) through [DomainCaps](https://github.com/libvirt/libvirt-go-xml/blob/master/domain_capabilities.go)  ?

Acedus commented:
> @Acedus In general this PR look like a great improvement to kubevirt, i'll give a deeper look soon.
> 
> Have you considered about doing the same in a follow up PR for [getDomCapabilities](https://github.com/kubevirt/kubevirt/blob/2513c96e5e545874219914d0cd5038b3d3a6e8b5/pkg/virt-handler/node-labeller/cpu_plugin.go#L161) through [DomainCaps](https://github.com/libvirt/libvirt-go-xml/blob/master/domain_capabilities.go) ?

Indeed, I still need to submit the design proposal for the overall change, but we're currently tracking it in https://github.com/kubevirt/kubevirt/issues/12727.

Acedus commented:
/test pull-kubevirt-e2e-k8s-1.30-sig-compute-serial
/test pull-kubevirt-unit-test-s390x
/test pull-kubevirt-e2e-k8s-1.31-sig-compute
/test pull-kubevirt-e2e-k8s-1.31-sig-operator

Barakmor1 commented on pkg/virt-handler/options_test.go:
```diff
@@ -0,0 +1,176 @@
+}
+
+func withCellDistances(distance int) option {
+	return func(caps *libvirtxml.Caps) {
+		for i := range caps.Host.NUMA.Cells.Cells {
+			caps.Host.NUMA.Cells.Cells[i].Distances = &libvirtxml.CapsHostNUMADistances{
+				Siblings: make([]libvirtxml.CapsHostNUMASibling, len(caps.Host.NUMA.Cells.Cells)),
+			}
+			for j := range caps.Host.NUMA.Cells.Cells {
+				// Doesn't mimic standard NUMA behavior, no differing latency
+				caps.Host.NUMA.Cells.Cells[i].Distances.Siblings[j] = libvirtxml.CapsHostNUMASibling{
+					ID:    j,
+					Value: distance,
+				}
+			}
+		}
+	}
+}
```
same here

iholder101 replied:
I agree. Also see the comment here https://github.com/kubevirt/kubevirt/pull/12652#discussion_r1743340868

Acedus replied:
Done.

Barakmor1 commented on pkg/virt-handler/options_test.go:
```diff
@@ -0,0 +1,176 @@
+	}
+}
+
+func withCellPages(pageCount int, count uint64, unit string) option {
+	return func(caps *libvirtxml.Caps) {
+		for i := range caps.Host.NUMA.Cells.Cells {
+			caps.Host.NUMA.Cells.Cells[i].PageInfo = make([]libvirtxml.CapsHostNUMAPageInfo, pageCount)
+			for j := range pageCount {
+				caps.Host.NUMA.Cells.Cells[i].PageInfo[j] = libvirtxml.CapsHostNUMAPageInfo{
+					Unit:  unit,
+					Count: count,
+					// Similar growth to common hugepages (4KiB, 2MiB. 1GiB)
+					Size: 4 << (9 * j),
+				}
+			}
+		}
+	}
+}
```
same here

Acedus replied:
Done.

Barakmor1 commented on pkg/virt-handler/options_test.go:
```diff
@@ -0,0 +1,176 @@
+	}
+}
+
+func withCellCpus(count int) option {
+	return func(caps *libvirtxml.Caps) {
+		for i := range caps.Host.NUMA.Cells.Cells {
+			caps.Host.NUMA.Cells.Cells[i].CPUS = &libvirtxml.CapsHostNUMACPUs{
+				CPUs: make([]libvirtxml.CapsHostNUMACPU, count),
+			}
+			for j := 0; j < count; j++ {
+				// Siblings alternate (e.g. {0,3} {1,2} {2,1} {3,0})
+				currCpu := j + (i * count)
+				caps.Host.NUMA.Cells.Cells[i].CPUS.CPUs[j] = libvirtxml.CapsHostNUMACPU{
+					ID:       currCpu,
+					Siblings: fmt.Sprintf("%d,%d", currCpu, count*(i+1)-j-1),
+				}
+			}
+
+		}
+	}
+}
```
Personally, I find hardcoded values easier to follow. The canonical loops in the tests seem a bit complex to me, and I feel we might be overcomplicating things.

Acedus replied:
Done.

Acedus commented on cmd/virt-handler/virt-handler.go:
```diff
@@ -124,6 +123,9 @@ const (
 	defaultTlsKeyFilePath     = "/etc/virt-handler/servercertificates/tls.key"
+
+	// Node capabilities path for exporting libvirt capabilities
+	nodeCapabilitiesVolumePath = "/var/lib/kubevirt-node-capabilities/"
```
Done.

Acedus commented on pkg/virt-operator/resource/generate/components/daemonsets.go:
```diff
@@ -123,15 +125,15 @@ func NewHandlerDaemonSet(namespace, repository, imagePrefix, version, launcherVe
-					Name:      "node-labeller",
-					MountPath: nodeLabellerVolumePath,
+					Name:      "node-capabilities",
+					MountPath: nodeCapabilitiesVolumePath,
```
Done.

Acedus commented:
[Change](https://github.com/kubevirt/kubevirt/compare/11720a3a7c934222982cffaa27eb6e8d6c76c70f..dbf8218b12d87d429325d4a78f41c9ac212eca4f) - Addressed @Barakmor1 and @iholder101 review comments.

iholder101 commented on pkg/virt-handler/options_test.go:
```diff
@@ -0,0 +1,175 @@
+	"libvirt.org/go/libvirtxml"
+)
+
+const memorySize = 16256896
```
Can you explain why this specific number? I suggest either adding a comment that clarifies that, or even better, assign it to the actual calculation that's taking place (i.e. `const memorySize = 2 * 1024 * 1024`)

Acedus replied:
Arbitrary, changed.

iholder101 commented on pkg/virt-handler/options_test.go:
```diff
@@ -0,0 +1,175 @@
+	}
+}
+
+func withCellPages(pageCount int, count uint64, unit string) option {
```
I think `count` is confusing. What's the difference between `count` and `pageCount`?

Acedus replied:
`pageCount` reflects the number of page size options per cell, count is the amount of pages of each page option (e.g. 1000 pages of size 4KiB). I think I can change it to `count` and `pageAmount` respectively to conform to the established convention of count == amount of objects.

Acedus commented on pkg/virt-handler/options_test.go:
```diff
@@ -0,0 +1,175 @@
+const memorySize = 16256896
+const pageAmount = 1000
+const cellDistance = 10
+const unit = "KiB"
```
I agree, done.

Acedus commented on pkg/virt-handler/options_test.go:
```diff
@@ -0,0 +1,175 @@
+
+const memorySize = 16256896
+const pageAmount = 1000
+const cellDistance = 10
```
Done.

Acedus commented:
[Change](https://github.com/kubevirt/kubevirt/compare/dbf8218b12d87d429325d4a78f41c9ac212eca4f..39b5b8a9c3997e607e97725c1101d08bdace5d86) - Addressed @iholder101 review comments.

orelmisan commented on pkg/virt-handler/options_test.go:
```diff
@@ -0,0 +1,171 @@
+package virthandler
```
nit: Please add the license comment.

Acedus replied:
Done.

orelmisan commented on pkg/virt-handler/options_test.go:
```diff
@@ -0,0 +1,171 @@
+var _ = Describe("Parsing VMI Options", func() {
+	Context("virt-handler VM processes VMI options during update", func() {
+		DescribeTable("it should convert api capabilities to cmdv1 topology", func(cellCount, pageCount, cpuCount int) {
+			caps := newLibvirtCapsTopology(
```
Please consider using a struct literal here instead of a builder - in order to make the input more explicit.

orelmisan commented on pkg/virt-handler/options_test.go:
```diff
@@ -0,0 +1,171 @@
+				withCellCpus(cpuCount),
+			)
+			topology := caps.Host.Topology
+			expectedTopology := capabilitiesToTopology(caps)
```
Please consider using a struct literal here instead of a function call - in order to make the output more explicit.

orelmisan commented on pkg/virt-handler/options_test.go:
```diff
@@ -0,0 +1,171 @@
+				Expect(expectedTopology.NumaCells[i].Pages).To(ConsistOf(matchPages(cell.Pages)))
+				Expect(expectedTopology.NumaCells[i].Distances).To(ConsistOf(matchDistances(cell.Distances.Sibling)))
+				Expect(expectedTopology.NumaCells[i].Cpus).To(ConsistOf(matchCPUs(cell.Cpus.CPU)))
+			}
```
Please consider asserting the actual value is equal to the expected:
```go
Expect(actual).To(Equal(expected))
```

orelmisan commented on pkg/virt-handler/node-labeller/node_labeller.go:
```diff
@@ -340,3 +340,10 @@ func (n *NodeLabeller) alertIfHostModelIsObsolete(originalNode *v1.Node, hostMod
 	return nil
 }
+
+func (n *NodeLabeller) GetTSCCounter() *libvirtxml.CapsHostCPUCounter {
```
Should this function be exported?

Acedus replied:
This method  is currently exported so that the `cpu_plugin_test.go` is able to make use of it during its test `should return correct cpu models, features and tsc freqnency`, the test is removed entirely in the third commit and the `GetTSCCounter()` method changes and becomes unexported.

orelmisan commented on pkg/virt-handler/node-labeller/cpu_plugin_test.go:
```diff
@@ -56,38 +57,36 @@ var _ = Describe("Node-labeller config", func() {
 			logger:                  log.DefaultLogger(),
 			volumePath:              "testdata",
 			domCapabilitiesFileName: "virsh_domcapabilities.xml",
+			cpuCounter:              cpuCounter,
```
Since this test isn't concerned about the CPU counter, please consider passing `nil` here.

Acedus replied:
Good catch, done.

orelmisan commented on pkg/virt-handler/node-labeller/node_labeller.go:
```diff
@@ -110,6 +111,10 @@ func (n *NodeLabeller) Run(threadiness int, stop chan struct{}) {
 
 	n.logger.Infof("node-labeller is running")
 
+	if !n.hasTSCCounter() {
+		n.logger.Error("failed to get tsc cpu frequency, will continue without the tsc frequency label")
+	}
```
Since this is an optimization, please consider introducing it in a follow-up commit:
1. It won't be lost in the bigger change.
2. It would have reasoning.

Barakmor1 replied:
I think that this was actually there before the PR have a look in `pkg/virt-handler/node-labeller/node_labeller.go` removed code

Acedus replied:
It was, but I agree with @orelmisan that it would be a good idea to add a comment about it moving out of `prepareLabels` to avoid redundant log messages.

orelmisan commented on pkg/virt-handler/node-labeller/node_labeller_test.go:
```diff
@@ -162,6 +170,17 @@ var _ = Describe("Node-labeller ", func() {
 		))
 	})
 
+	It("should add cpu tsc labels if cpu counter exists and its name is tsc", func() {
```
Please consider converting this test case into a table:
1. Test that no labels were added when there is no tsc counter.
2. Test the `v1.CPUTimerLabel+"tsc-scalable"` label when `Scaling` is true.
3. Same when its false.

orelmisan replied:
On a second thought, the first TC will not play nicely with the other two in a single table.

Acedus replied:
In that case, I'll be proceeding with 2 and 3.

orelmisan replied:
You could have an "It()" for the first test case, and a table for the other two.

Acedus replied:
Done.

orelmisan commented on cmd/virt-handler/virt-handler.go:
```diff
@@ -307,8 +307,9 @@ func (app *virtHandlerApp) Run() {
 	defer close(stop)
 	var capabilities libvirtxml.Caps
 	var hostCpuModel string
+	nodeCapabilitiesVolumePath := "/var/lib/kubevirt-node-capabilities/"
```
I personally feel that this PR is already quite big and touches multiple files.
Please consider introducing this change in a follow-up, I don't think its a must in this PR.

Acedus replied:
I don't mind postponing this change to a separate PR.
Any issue with that @iholder101?

Acedus replied:
After deliberation on the matter I've decided to drop the commit for the time being, will postpone it to a separate PR.

orelmisan reviewed (commented):
Thank you for the changes @Acedus.
Please see the inline comments.

Acedus commented:
[Change](https://github.com/kubevirt/kubevirt/compare/39b5b8a9c3997e607e97725c1101d08bdace5d86..e3675523d5b98a48a3a8690c3cd5cba3935270bd) - Addressed @orelmisan review comments.

[Change](https://github.com/kubevirt/kubevirt/compare/e3675523d5b98a48a3a8690c3cd5cba3935270bd..720c88a3bdaebc0065ed0cb6e00bdb486b70fd15) Added commit to introduce the movement of the TSC counter assertion from the `prepareLabels` loop to `Run`.

Acedus commented:
/hold

Waiting on #12761.

Acedus commented:
[Change](https://github.com/kubevirt/kubevirt/compare/a76c7d4d4318c8a0402fbb07b47774d941e31130..555891d4087dac7258552fd7a9b5699aab1c316a) - Removed commit that introduced `options_test.go` in favor of https://github.com/kubevirt/kubevirt/pull/12761 and adjusted the tests to use libvirtxml.

[Change](https://github.com/kubevirt/kubevirt/compare/a76c7d4d4318c8a0402fbb07b47774d941e31130..555891d4087dac7258552fd7a9b5699aab1c316a) - Rebased to include changes made in https://github.com/kubevirt/kubevirt/pull/12761.

Acedus commented:
/unhold

Acedus commented:
/retest

Acedus commented:
[Change](https://github.com/kubevirt/kubevirt/compare/555891d4087dac7258552fd7a9b5699aab1c316a..d54b2234d47cdfe84b63d8edaaa22bc79e433739) - Bazel source discrepancy.

Acedus commented:
/retest

Acedus commented:
[Change](https://github.com/kubevirt/kubevirt/compare/d54b2234d47cdfe84b63d8edaaa22bc79e433739..259f1b21f39ba54536a244c7527d613e2591ddec) - Dropped commit that renames `node-labeller.sh` to a more fitting name.

orelmisan commented on pkg/virt-handler/node-labeller/node_labeller_test.go:
```diff
@@ -162,6 +169,34 @@ var _ = Describe("Node-labeller ", func() {
+			HaveKeyWithValue(v1.CPUTimerLabel+"tsc-scalable", result),
+		))
+	},
+		Entry("scaling is set to not", "no", "false"),
```
```suggestion
		Entry("scaling is set to no", "no", "false"),
```

Acedus replied:
Done.

orelmisan commented on pkg/virt-handler/node-labeller/node_labeller_test.go:
```diff
@@ -162,6 +169,34 @@ var _ = Describe("Node-labeller ", func() {
+			Not(HaveKey(v1.CPUTimerLabel+"tsc-frequency")),
+			Not(HaveKey(v1.CPUTimerLabel+"tsc-scalable")),
+		))
+
```
nit: Redundant newline.

Acedus replied:
Done.

orelmisan reviewed (approved):
Thank you @Acedus, this is a great step forward!
@iholder101, @Barakmor1 is there anything else?
/hold

Acedus commented:
[Change](https://github.com/kubevirt/kubevirt/compare/259f1b21f39ba54536a244c7527d613e2591ddec..92bc283e6c2111b8d714239535ead8ef2ec90699) - Addressed @orelmisan's comments.

orelmisan commented:
/test pull-kubevirt-unit-test

orelmisan commented:
/hold cancel

orelmisan commented:
/retest-required

vamsikrishna-siddu commented:
Hi @Acedus , After your PR got merged. The kubevirt is not working for s390x Architecture. The Reason is when we install the kubevirt with the latest main, the virt-handler is going to Crash loop back off state. Here are the logs.

Panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x81e68698]

```
goroutine 82 [running]:
kubevirt.io/kubevirt/pkg/virt-handler/node-labeller.(*NodeLabeller).hasTSCCounter(0xc0005a3340)
	pkg/virt-handler/node-labeller/node_labeller.go:338 +0xe8
kubevirt.io/kubevirt/pkg/virt-handler/node-labeller.(*NodeLabeller).Run(0xc0005a3340, 0xa, 0xc0000b5860)
	pkg/virt-handler/node-labeller/node_labeller.go:114 +0xea
created by main.(*virtHandlerApp).Run in goroutine 1
	cmd/virt-handler/virt-handler.go:335 +0x1796
```
https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-handler/node-labeller/node_labeller.go#L338
n.CpuCounter is coming as nil for s390x. This is because TSC is not supported on s390x.

Can you please fix this?

Thanks

Acedus commented:
> Hi @Acedus , After your PR got merged. The kubevirt is not working for s390x Architecture. The Reason is when we install the kubevirt with the latest main, the virt-handler is going to Crash loop back off state. Here are the logs.
> 
> Panic: runtime error: invalid memory address or nil pointer dereference [signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x81e68698]
> 
> ```
> goroutine 82 [running]:
> kubevirt.io/kubevirt/pkg/virt-handler/node-labeller.(*NodeLabeller).hasTSCCounter(0xc0005a3340)
> 	pkg/virt-handler/node-labeller/node_labeller.go:338 +0xe8
> kubevirt.io/kubevirt/pkg/virt-handler/node-labeller.(*NodeLabeller).Run(0xc0005a3340, 0xa, 0xc0000b5860)
> 	pkg/virt-handler/node-labeller/node_labeller.go:114 +0xea
> created by main.(*virtHandlerApp).Run in goroutine 1
> 	cmd/virt-handler/virt-handler.go:335 +0x1796
> ```
> 
> https://github.com/kubevirt/kubevirt/blob/main/pkg/virt-handler/node-labeller/node_labeller.go#L338 n.CpuCounter is coming as nil for s390x. This is because TSC is not supported on s390x.
> 
> Can you please fix this?
> 
> Thanks

Hi @vamsikrishna-siddu, thank you for raising this.
This is quite alarming, I'll be reverting this patch until it is resolved.

orelmisan commented:
Thank you for raising it @vamsikrishna-siddu.
@Acedus, I don't think we should to revert the whole PR - lets just fix the bug.
